### PR TITLE
stack implementation

### PR DESCRIPTION
**Pull request by KaveriMoger**
please check and do the changes if necessary

### CODE REVIEW

File: stack/stack.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      @store[@top] = element
+    end
+    @top += 1
+  end
+
+  def pop
+    if empty? or @store.nil?
+      p "empty"
+    else
+      @store.pop
```
Comment (Sushanthbc) [inline stack/stack.rb]: Let's not use ```pop``` questions is how are we going about it?

File: stack/stack.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    end
+    @top -= 1
+  end
+
+
+  private
+    def empty?
```
Comment (Sushanthbc) [inline stack/stack.rb]: spaces are required here?

File: stack/stack.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
+class Stack
+  def initialize(size)
+    @size = size
+    @store = Array.new(@size)
+    @top = -1
+  end
+
+  def push(element)
+    if full?
```
Comment (Sushanthbc) [inline stack/stack.rb]: why don't we use guarded clause ``` raise "overflow error" if full? ```
Also, we raise error are we handling it?  we should build some kind of logger to handle and print information.

File: stack/tc_stack.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    size= Stack.new(2)
+    item = size.push(6)
+    item = size.push(7)
+    assert_equal(1,item)
+  end
+
+  def test_check_overflow
+    size= Stack.new(1)
+    item = size.push(6)
+    assert_raise RuntimeError do
```
Comment (Sushanthbc) [inline stack/tc_stack.rb]: Let's move away from this RuntimeError, we send user message. `return "stack is overflow"`

File: stack/tc_stack.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      item = size.push(9)
+    end
+  end
+
+  def test_remove_item_to_stack
+    size= Stack.new(2)
```
Comment (Sushanthbc) [inline stack/tc_stack.rb]: okay, let's give space before `size = `

File: stack/tc_stack.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    item = size.push(6)
+    item = size.push(7)
+    size.pop
+    assert_equal(1,item)
+  end
+
+  def test_check_underflow
+    size= Stack.new(1)
```
Comment (Sushanthbc) [inline stack/tc_stack.rb]: okay, let's give space before `size = `

File: stack/stack.rb [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        popped = @store[@top]
+        @store[@top] = nil
+        @top = @top.pred
+        popped
+      end
+    end
+
+    def remove_element(element)
+      if empty?
+        raise "underflow error"
```
Comment (Sushanthbc) [inline stack/stack.rb]: lets have `return here`

File: stack/tc_stack.rb [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    item = size.push(6)
+    size.pop
+    assert_raise RuntimeError do
```
Comment (Sushanthbc) [inline stack/tc_stack.rb]: if you are using `return` this `assert_raise` will change

File: stack/tc_stack.rb [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+      size.pop
+    end
+  end
+
+  def test_remove_selected_item
+    size= Stack.new(3)
+    item = size.push(6)
```
Comment (Sushanthbc) [inline stack/tc_stack.rb]: include elements in array and pushing will make more sense and looks clean :P

File: stack/stack.rb (hunk 2 above)
Comment (SmrutiSuman) [reply stack/stack.rb]: @KaveriMoger we should not override ruby build in functionalities(empty?) for a specific purpose, either we should define it for generic purpose or should rename it.

File: stack/stack.rb [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        raise "overflow error"
+      else
+        @top += 1
+        @store[@top] = element
+      end
+    end
+
+    def pop
+      if empty? or @store.nil?
+        raise "underflow error"
```
Comment (SmrutiSuman) [inline stack/stack.rb]: It is better if the message is capitalized.

File: stack/stack.rb [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+      else
+        deletion_index = nil
+        @store.each_with_index do |ele, idx|
+          if ele == element
+            deletion_index = idx
+            break
+          end
+        end
+        (@top - deletion_index + 1).times do |i|
```
Comment (SmrutiSuman) [inline stack/stack.rb]: No need of this loop, inside the same loop also we can do changes.Can pop inside the above loop.

**Pull request merged**